Enable bulk action by Jira filter

chandler05 opened the pull request:
## Purpose
Large amounts of requests like adding versions can end up being very tedious to do Individually. With a new reaction and command, the bot will provide a Jira filter will all of the tickets from specific requests in order to use a Bulk Action.
## Approach
Workflow:
1. Add the bulk reaction to any requests
2. Type '!jira bulk'
3. The bot will send a Jira link with a filter containing all tickets in the reacted to requests
4. The bot will remove the reactions
## Future work

SPGoding reviewed (commented):
From what I understand, the bulk emoji will result in showing the prompt and solving it after 10 seconds as there's no special check in `RequestEventHandler`.

Also, how about adding tickets IDs to an existing `Set` as soon as someone reacted an request with the `bulkEmoji`, so we're not iterating the whole internal channel when executing `!jira bulk` command?

chandler05 commented:
Wouldn't that stop the filters from being individual for each person?

SPGoding commented:
It's still possible to make a `Map` from user tags to `Set`s :eyes:


Awwwso, instead of asking the volunteer to execute `!jira bulk` manually to see the filter, how about adding it to a new field of internal embed in some format like `[bulk filter](https://bugs.mojang.com/jql@#%$^^%)` as soon as the volunteer reacted it with the `bulkEmoji`? This way, the `Set` (or `Map`) maintained in `RequestEventHandler` doesn't have to be exposed to another handler.

chandler05 commented:
Where would that be posted?

SPGoding commented:
![image](https://user-images.githubusercontent.com/15277496/101262322-60006400-3703-11eb-96ac-60aa99dbd3a2.png)

chandler05 commented:
The filter is supposed to extend over any request with the reaction. That confuses me as to how that would work. Plus, wouldn't that only work for one user?

SPGoding commented:
Yeah I just gave it another thought and it's quite difficult to maintain if we put the JQL link directly in the embed. So I guess we still have to go with the `!jira bulk` command 🤔

chandler05 commented:
I changed the code to make it so a map of users to bulk messages is stored instead of iterating through all of the messages in the channel each time. I also fixed the bulk emoji from triggering the prompt/resolution.

chandler05 commented:
Adding an emoji after '!jira bulk' should now resolve all of the requests in the bulk filter with that emoji

SPGoding commented on src/events/request/RequestBulkRemoveEventHandler.ts:
```diff
@@ -0,0 +1,27 @@
+			for ( let i = 0; i <= currentMessages.length; i++ ) {
+				const currentMessage = currentMessages[i];
+				if ( currentMessage === reaction.message ) {
+					BulkCommand.currentBulkReactions.delete( user );
+					BulkCommand.currentBulkReactions.set( user, currentMessages.splice( i, 1 ) );
```
`splice` returns the deleted elements, not the array itself. Also, `currentMessages` is a reference to the array which isn't reassigned anywhere, so all we need to do is a simple
```suggestion
					currentMessages.splice( i, 1 );
```

SPGoding commented on src/commands/BulkCommand.ts:
```diff
@@ -0,0 +1,82 @@
+
+		try {
+			const bulkMessages = BulkCommand.currentBulkReactions.get( message.author );
+			let originMessages: Message[];
```
This is never initialized.

SPGoding commented on src/commands/BulkCommand.ts:
```diff
@@ -0,0 +1,82 @@
+			const customEmoji = /^<a?:(\w+):(\d+)>/;
+			const unicodeEmoji = emojiRegex();
+
+			if ( customEmoji.test( args ) || unicodeEmoji.test( args ) ) {
+				rawEmoji = args;
+				const emojiMatch = customEmoji.exec( args );
+				if ( emojiMatch ) {
+					rawEmoji = emojiMatch[2];
+				}
```
The logic here looks tricky. I'd suggest using

```suggestion
			if ( unicodeEmoji.test( args ) ) {
				rawEmoji = args;
			} else if ( customEmoji.test( args ) ) {
				rawEmoji = customEmoji.exec( args )[2];
```

SPGoding commented on src/commands/BulkCommand.ts:
```diff
@@ -0,0 +1,82 @@
+		try {
+			const bulkMessages = BulkCommand.currentBulkReactions.get( message.author );
+			let originMessages: Message[];
+			for ( const bulk of bulkMessages ) {
```
`bulkMessages` could be `undefined`.

SPGoding commented on src/commands/BulkCommand.ts:
```diff
@@ -0,0 +1,82 @@
+		return true;
+	}
+
+	private getTickets( content: string ): string[] {
+		let ticketMatch: RegExpExecArray;
+		const regex = MentionCommand.getTicketIdRegex();
+		const ticketMatches: string[] = [];
+		while ( ( ticketMatch = regex.exec( content ) ) !== null ) {
+			ticketMatches.push( ticketMatch[1] );
+		}
+		return ticketMatches;
+	}
```
I'd suggest moving both this and the one in `RequestEventHandler` to `RequestsUtil`.

SPGoding commented on src/commands/BulkCommand.ts:
```diff
@@ -0,0 +1,82 @@
+			}
+		}
+
+		let ticketKeys: string[];
```
This is never initialized.

SPGoding commented on src/commands/BulkCommand.ts:
```diff
@@ -0,0 +1,82 @@
+				originMessages.push( await RequestsUtil.getOriginMessage( bulk ) );
+				await bulk.reactions.cache.get( BotConfig.request.bulkEmoji ).users.remove( message.author );
+			}
+			originMessages.forEach( origin => this.getTickets( origin.content ).forEach( ticket => ticketKeys.push( ticket ) ) );
```
The nested `forEach` here is confusing.
```suggestion
			originMessages.forEach( origin => ticketKeys.push( ...this.getTickets( origin.content ) ) );
```

SPGoding commented on src/commands/BulkCommand.ts:
```diff
@@ -0,0 +1,82 @@
+			originMessages.forEach( origin => this.getTickets( origin.content ).forEach( ticket => ticketKeys.push( ticket ) ) );
+			firstMentioned = ticketKeys[0];
+			if ( rawEmoji ) {
+				bulkMessages.forEach( resolvable => TaskScheduler.addOneTimeMessageTask(
+					resolvable,
+					new ResolveRequestMessageTask( rawEmoji, message.author ),
+					BotConfig.request.resolveDelay || 0
+				) );
```
I'd suggest delegating this operation to `RequestResolveEventHandler` instead of copying this piece of code here.

chandler05 replied:
The only problem is there is no actual message reaction to handle at this point, just the emoji

SPGoding commented on src/commands/BulkCommand.ts:
```diff
@@ -0,0 +1,82 @@
+			return false;
+		}
+
+		return true;
```
I think it would be fine to delete the `!jira bulk` command somewhere. And also delete the bot's own filter message after some time.

chandler05 commented:
The feature is working very well, but for some reason it seems to ocassionally miss some of the bulk emoji reactions in the internal channels
